src/utils/useSessionState.ts

```diff
@@ -0,0 +1,79 @@
+import { Dispatch, SetStateAction, useCallback, useEffect, useMemo, useRef, useState } from "react";
+
+import { ADDON_ID } from "../constants";
+
+const debounce = (callback: (...args: any[]) => unknown, wait: number) => {
+  let timeoutId: number;
+  const cancel = () => {
+    window.clearTimeout(timeoutId);
+    timeoutId = 0;
+  };
+  const debounced = (...args: any[]) => {
+    if (timeoutId) cancel();
+    else callback(...args); // Leading edge call
+
+    timeoutId = window.setTimeout(() => {
+      callback(...args); // Trailing edge call
+      timeoutId = 0;
+    }, wait);
+  };
+  debounced.cancel = cancel;
+  return debounced;
+};
+
+export function useSessionState<S>(
```

Comment:
This hook stores values into the sessionStorage AND has a local state in line 28 to preserve the current value. The problem with this approach is that if we have two components, which both use, let's say, `useSessionState('a,' initialState), and one of the components update the session state, the other one will not be notified since there aren't any session storage listeners in place. Indeed, my recommendation would be not to reinvent the wheel but to rely on existing solutions like `https://github.com/juliencrn/usehooks-ts/blob/master/packages/usehooks-ts/src/useSessionStorage/useSessionStorage.ts`

Reply:
Yeah, we should use `useSharedState` for that. I agree though that it's a footgun.
Generally I'm not keen on adding extra dependencies if I can avoid it. I'll take a look at this one though.